File: mainActivity/src/main/java/com/sierrawireless/avphone/tools/MyPreference.java
package com.sierrawireless.avphone.tools;

import android.content.SharedPreferences;
import android.text.TextUtils;

import com.google.gson.Gson;
import com.sierrawireless.avphone.model.AvPhoneObject;

import java.util.ArrayList;
import java.util.Arrays;

public class MyPreference {
    private SharedPreferences preferences;

    public MyPreference(SharedPreferences prefences) {
        this.preferences = prefences;
    }

    /**
     * Get String value from SharedPreferences at 'key'. If key not found, return ""
     * @param key SharedPreferences key
     * @return String value at 'key' or "" (empty String) if key not found
     */
    public String getString(String key) {
        return preferences.getString(key, "");
    }

    /**
     * Get parsed ArrayList of String from SharedPreferences at 'key'
     * @param key SharedPreferences key
     * @return ArrayList of String
     */
    private ArrayList<String> getListString(String key) {
        return new ArrayList<>(Arrays.asList(TextUtils.split(preferences.getString(key, ""), "‚‗‚")));
    }

    public ArrayList<AvPhoneObject> getListObject(String key, Class<AvPhoneObject> mClass){
        Gson gson = new Gson();

        ArrayList<String> objStrings = getListString(key);
        ArrayList<AvPhoneObject> objects = new ArrayList<>();

        for(String jObjString : objStrings){
            AvPhoneObject value  = gson.fromJson(jObjString,  mClass);
            objects.add(value);
        }
        return objects;
    }


    /**
     * Get int value from SharedPreferences at 'key'. If key not found, return 'defaultValue'
     * @param key SharedPreferences key
     * @return int value at 'key' or 'defaultValue' if key not found
     */
    public int getInt(String key) {
        return preferences.getInt(key, 0);
    }

    /**
     * Put int value into SharedPreferences with 'key' and save
     * @param key SharedPreferences key
     * @param value int value to be added
     */
    public void putInt(String key, int value) {
        checkForNullKey(key);
        preferences.edit().putInt(key, value).apply();
    }
    /**
     * null keys would corrupt the shared pref file and make them unreadable this is a preventive measure
     * @param key pref key
     */
    private void checkForNullKey(String key){
        if (key == null){
            throw new NullPointerException();
        }
    }

    /**
     * Put ArrayList of String into SharedPreferences with 'key' and save
     * @param key SharedPreferences key
     * @param stringList ArrayList of String to be added
     */
    private void putListString(String key, ArrayList<String> stringList) {
        checkForNullKey(key);
        String[] myStringList = stringList.toArray(new String[stringList.size()]);
        preferences.edit().putString(key, TextUtils.join("‚‗‚", myStringList)).apply();
    }

    public void putListObject(String key, ArrayList<AvPhoneObject> objArray){
        checkForNullKey(key);
        Gson gson = new Gson();
        ArrayList<String> objStrings = new ArrayList<>();
        for(Object obj : objArray){
            objStrings.add(gson.toJson(obj));
        }
        putListString(key, objStrings);
    }
}

File: mainActivity/src/main/java/com/sierrawireless/avphone/tools/MyPreference.kt
package com.sierrawireless.avphone.tools

import android.content.SharedPreferences
import android.text.TextUtils
import com.google.gson.Gson
import com.sierrawireless.avphone.model.AvPhoneObject
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

class MyPreference(private val preferences: SharedPreferences) {

    /**
     * Get String value from SharedPreferences at 'key'. If key not found, return ""
     * @param key SharedPreferences key
     * @return String value at 'key' or "" (empty String) if key not found
     */
    fun getString(key: String): String {


        return preferences.getString(key, "")!!
    }

    /**
     * Get parsed ArrayList of String from SharedPreferences at 'key'
     * @param key SharedPreferences key
     * @return ArrayList of String
     */
    private fun getListString(key: String): ArrayList<String> {
        return ArrayList(listOf(*TextUtils.split(preferences.getString(key, ""), "‚‗‚")))
    }

    fun getListObject(key: String, mClass: Class<AvPhoneObject>): CopyOnWriteArrayList<AvPhoneObject> {
        val gson = Gson()

        val objStrings = getListString(key)
        return objStrings.mapTo(CopyOnWriteArrayList()) { gson.fromJson(it, mClass) }

    }


    /**
     * Get int value from SharedPreferences at 'key'. If key not found, return 'defaultValue'
     * @param key SharedPreferences key
     * @return int value at 'key' or 'defaultValue' if key not found
     */
    fun getInt(key: String): Int {
        return preferences.getInt(key, 0)
    }

    /**
     * Put int value into SharedPreferences with 'key' and save
     * @param key SharedPreferences key
     * @param value int value to be added
     */
    fun putInt(key: String, value: Int) {
        checkForNullKey(key)
        preferences.edit().putInt(key, value).commit()
    }

    /**
     * null keys would corrupt the shared pref file and make them unreadable this is a preventive measure
     * @param key pref key
     */
    private fun checkForNullKey(key: String?) {
        if (key == null) {
            throw NullPointerException()
        }
    }

    /**
     * Put ArrayList of String into SharedPreferences with 'key' and save
     * @param key SharedPreferences key
     * @param stringList ArrayList of String to be added
     */
    private fun putListString(key: String, stringList: CopyOnWriteArrayList<String>) {
        checkForNullKey(key)
        val myStringList = stringList.toTypedArray()
        preferences.edit().putString(key, TextUtils.join("‚‗‚", myStringList)).commit()
    }

    fun putListObject(key: String, objArray: CopyOnWriteArrayList<AvPhoneObject>) {
        checkForNullKey(key)
        val gson = Gson()
        val objStrings = objArray.mapTo(CopyOnWriteArrayList<String>()) { gson.toJson(it) }
        putListString(key, objStrings)
    }
}
